### ISSUE TITLE
Automate CHANGELOG generation

### ISSUE THREAD

**Issue by yardnsm** [wix/wix-style-react#2682]
Currently, our [CHANGELOG](https://github.com/wix/wix-style-react/blob/master/CHANGELOG.md) file is written by hand. after every merged PR we should manually update it, but sometimes we may forgot to. When preparing a new release, we also need to manually update the version and possibly add more entries that.

This can be tedious sometimes, and can be automated. I suggest using [`lerna-changelog`](https://github.com/lerna/lerna-changelog) for this, it can be done in the CI while we the `release` script is being ran. Using `lerna-changelog` will force us to use Git tags (which can also be created in CI) and adding labels to our PRs, so +1 for that.

The [Yoshi](https://github.com/wix/yoshi/blob/master/CHANGELOG.md) team is using it (and they also has emojis).

Wdyt?

@shlomitc @hepiyellow @mykas @argshook @nitayneeman 

Also, since it's format is a bit different than ours, we can deprecate the current CHANGELOG (which is very long...).

**Comment by hepiyellow** [comment]
I tried once looking into a few auto changelog tools.

Can you do a little demo next weekly about Lerna-changelog:
- What is the normal flow
- Is it incremental? or it generates always the whole change log history?
- How do we manually : 
  - Add changelog entries
  - ignore changelog entries (ignore commits from showing up in the changelog, like internal refactors and such)


etc...

**Comment by argshook** [comment]
when we first introduced changelog i was looking into how to automate them but came to a conclusion, that it's not worth it:

1. at the time i didn't find tool good enough for the purpose. what i found were very opinionated, had strange formats that were hard (or impossible) to change, or use things like ruby, which would be a really unnecessary burden in our js land. we could write some tool on our own but we chose to keep it simple and do it manually.
2. automated changelogs look like git log but in markdown. is that really useful? git log is already available on github.
3. changelog is meant to be read by human and so it should be written by human. what i mean here is that it can't be just a translation of git log. it needs to be filtered, sometimes links or descriptions added, sometimes not. this is not easy to automate.
4. `lerna-changelog` might be useful for monorepos (wix-ui) but since wix-style-react is not one, generated changelog wouldn't look too different from git log
5. we need to be able to edit old entries without being afraid that automated tool might not understand some syntax or remove edits entirely.
6. 
  > The Yoshi team is using it (and they also has emojis).

  they have monorepo and automating helps keep track of multiple modules. but even then i'd argue that there's a lot of noise. e.g. there are mentions about CI fixes, or prettier for markdown files, or improving some tests. all of these are useless info for consumer who just want to know what changes affect them, not some internal updates.

  > and they also has emojis

  maybe im getting old but that's not a plus at all to me.

7. 
  > Also, since it's format is a bit different than ours, we can deprecate the current CHANGELOG (which is very long...).

  our changelog format is based on https://keepachangelog.com which imo is a good one, not sure how we'd benefit if we were to change it.

  so what if our changelog is long? what's the problem? i think its easier for reader to look at one file and not jump around. we can move old entries (like from 1.0.0) to separate file, but then we need to have a link, keep separate file somewhere, don't forget to update link after moving it etc.


so in conclusion, manual changelog for our case is good enough imo.

---

sorry for wall of text, i didn't type much during holidays :)

**Comment by yardnsm** [comment]
@argshook 

1. `lerna-changelog` is written in TypeScript, so it would integrated fairly nicly with our environment.
1. Well, no. It scans the merged PR since the last tag, and only adds PRs with specific tags to the CHANGELOG.
1. Some as 2, and yeah, adding additional content may be problematic and we should look into it if we'll decide to proceed.
1. It's not target directly to monorepos, it should work the same for non-monorepos.
1. I agree, and that's probably the biggest downside to using this tool. We would probably have to deprecate our current changelog if we'd want to use it...
1. We can decide to not include infra stuff to the changelog. The Yoshi team simply has a "Polish" tag for PRs, like [here](https://github.com/wix/yoshi/pull/786).

> so what if our changelog is long? what's the problem?

It's not *the* problem I was talking about. But when using a tool like `lerna-changelog` we can force ourselves to using labels in our PRs, write more concise PR titles, and using git tags. It does have some downsides, like deprecating oue current changelog and not having the ability to manually add / change entires (although they may be a solution to this?).

I'm not saying its the perfect solution, but IMHO we should consider it and try it.

**Comment by gilmeir** [comment]
@argshook Referring to the points you brought up here:
2. It is useful IMO, exactly because the commits data is available, and we can generate a draft changelog item for every commit, and spare the manual work.
3. I believe if we go with the approach of auto-changelog creating only a draft, that's also valuable.
5. Of course. The tool should not change the changelog history, it should only suggest a draft for the version being released, and can also enforce/validate a changelog entry for every commit (or define exceptions somehow).
6. The release manager can easily remove those

I encourage us to start using tags for releases, it will be much easier to manage bookkeeping between versions.

**Comment by shlomitc** [comment]
This has been discussed several times and considered a good option, just not prioritized. I'll be happy if someone would take this in the future, but not prioritized, so closing it to keep a cleaner backlog.
